### PR TITLE
Cert auth

### PR DESCRIPTION
**Pull request by jhazentia**
Replaces OAuth/JWT authentication with certificate based authentication for authentication to sensor and to callback URL
- Browsable api timeout, logout doesn't work, always logged in
- Admin interface still uses username/password
- Updated README with some updates specific to this pull request and some more general updates
- Added ADDITIONAL_USER_NAMES and ADDITIONAL_USER_PASSWORD environment variables to programmatically create additional users when scos-sensor starts. ADDITIONAL_USER_NAMES can be comma separated. ADDITIONAL_USER_PASSWORD is not required even when using ADDITIONAL_USER_NAMES. However, without a password, the user will not be able to log in with session authentication or to the admin interface. However, certificate and token authentication will still work.
- Added ADMIN_NAME environment variable to control username for admin account
- Removed environment variables related to OAuth: CLIENT_ID, CLIENT_SECRET, OAUTH_TOKEN_URL, PATH_TO_JWT_PUBLIC_KEY, USER_NAME, PASSWORD
- Removed SENTRY_DSN environment variable
- Modified NGINX config for ssl client verification and to pass the certificate DN in request header
- Removed restore_fixture.sh
- Removed role based permissions; require all users to be superusers
- Added tests

### CODE REVIEW

**Review by dboulware — CHANGES_REQUESTED** [review]
This looks good. Only some minor comments to address.

File: README.md [hunk 1]
```diff
@@ -507,39 +525,34 @@ openssl pkcs12 -export -out client.pfx -inkey client.key -in client.pem -certfil
...
+[Nginx documentation](https://nginx.org/en/docs/http/ngx_http_ssl_module.html) for
+more information about configuring Nginx.
+
+To disable client certificate authentication, comment out the
+following in [nginx/conf.template](nginx/conf.template):
 
-```bash
-openssl pkey -inform PEM -outform PEM -in jwt.pem -out jwt_private_key.pem
+```text
+ssl_client_certificate /etc/ssl/certs/ca.crt;
```
[the remark this answers is not in the document]
Comment (jhazentia) [reply README.md:547]: Per our discussion, we decided to leave as one template. Ansible can do string replacement.

File: src/authentication/auth.py [hunk 2]
```diff
@@ -1,86 +1,49 @@
...
+        cert_dn = request.headers.get("X-Ssl-Client-Dn")
+        if cert_dn:
+            user_model = get_user_model()
+            try:
+                cn = get_cn_from_dn(cert_dn)
+                user = user_model.objects.get(username=cn)
+                user.last_login = datetime.datetime.now()
+                user.save()
+            except user_model.DoesNotExist:
+                raise exceptions.AuthenticationFailed("No matching username found!")
```
[the remark this answers is not in the document]
Comment (jhazentia) [reply src/authentication/auth.py:34]: Fixed in 3590675

File: scripts/create_superuser.py [hunk 3]
```diff
@@ -13,24 +13,60 @@
...
+        "Retreived additional user password from environment variable ADDITIONAL_USER_PASSWORD"
+    )
+except KeyError:
+    print("Not creating any additonal users.")
+
+
+if additional_user_names != "" and additional_user_password != "":
+    if "," in additional_user_names:
+        for additional_user_name in additional_user_names.split(","):
+            add_user(additional_user_name, additional_user_password)
```
[the remark this answers is not in the document]
Comment (jhazentia) [reply scripts/create_superuser.py]: Fixed in 3590675

File: README.md [hunk 4]
```diff
@@ -507,39 +525,34 @@ openssl pkcs12 -export -out client.pfx -inkey client.key -in client.pem -certfil
...
-
-###### Step 2: Extract Public Key
-
-```bash
-openssl rsa -in jwt.pem -outform PEM -pubout -out jwt_public_key.pem
-```
+###### Configure scos-sensor
 
-###### Step 3: Extract Private Key
+The Nginx web server is configured by default to require client certificates (mutual
```
[the remark this answers is not in the document]
Comment (jhazentia) [reply README.md]: I made some changes. Please check if new version is good (commit e9b2f30)

File: nginx/conf.template [hunk 5]
```diff
@@ -49,11 +49,17 @@ server {
 
     # Pass off requests to Gunicorn
     location @proxy_to_wsgi_server {
+        # if ($ssl_client_verify != SUCCESS) {
```
[the remark this answers is not in the document]
Comment (jhazentia) [reply nginx/conf.template]: Tested and confirmed that this does not seem to make any difference. Removed in e9b2f30

File: README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (jhazentia) [reply README.md]: Removed `ssl_ocsp on ` setting from README Nginx configuration text blocks. This has not been tested and unsure if additional settings required.

**Review by dboulware — CHANGES_REQUESTED** [review]
Looks good. Just one minor request. Add to the log message the user that could not be found when unable to find the user.

**Review by dboulware — APPROVED** [review]

**Review by dboulware — CHANGES_REQUESTED** [review]
Looks good. Can you resolve the conflicts?

**Comment by jhazentia** [comment]
> Looks good. Can you resolve the conflicts?

Done

**Review by dboulware — APPROVED** [review]

**Pull request merged**